doc/RFC-MMP.llnl.04.txt

```diff
@@ -0,0 +1,913 @@
+using devices already part of another pool.
+
+=== ext3/ldiskfs ===
+
+The existing multi-mount protection (MMP) mechanism in ext3/ldiskfs is based on
+having a block in the filesystem (the MMP block) which is updated periodically
+(e.g., every 5 seconds) to indicate that a node that has mounted the filesystem
+is still alive. Using this heartbeat-like mechanism, it is possible for another
+node that is attempting to mount the filesystem to check (among other things)
+whether this block is changing and therefore determine if it should be safe to
+mount the filesystem or not.
+
+For ZFS, there are several new considerations that must be taken into account,
+one of them being the fact that a ZFS pool can have multiple devices.  Not all
+of them may be accessible at a given point in time.  In addtion, there are
+scenarios such as adding a device to an existing pool, or use of a hot spare,
+where different nodes must not attempt to use the same device at the same time.
+
+Furthermore, we would like to take this opportunity to make a few improvements
+to the previous design.
+
+=== Multi-modifier protection struct (MMP struct) ===
+
+The MMP structure is a piece of data periodically written to stable storage
+that will allow a node that wants to open a pool in read/write mode to make an
+educated guess of whether another node also has the pool currently open.  It
+should contain at least the following data:
+
+struct mmp_struct {
+      uint64_t mmp_magic;
+      uint64_t mmp_pool_guid;
+      uint64_t mmp_open_id;
+      uint64_t mmp_seq;
+      uint32_t mmp_interval;
+      uint32_t mmp_delay;
+      zio_eck_t mmp_eck;
+
+/* Nice-to-have fields: */
+      char		mmp_nodename[64];
+};
+
+* mmp_magic:     Constant value used for sanity checking.
+* mmp_pool_guid: Pool guid.
+* mmp_open_id:   A random 64-bit value generated when opening the pool in
+		 read/write mode. A special constant value (MMP_CLEAN_ID)
+		 indicates the pool has been closed cleanly.
+* mmp_seq:       A sequence number. Starts at 0 when a pool is first opened
+		 and increases by at least 1, every mmp_interval period.
+		 Also increases by at least 1, at each spa_sync.
+* mmp_interval:  Fixed interval at which MMP blocks are being written
+		 (in milliseconds, usually 5000).  They are also written
+		 independently, based on pool change.
+* mmp_delay:	 Delay being experienced (in milliseconds) by CPU scheduling
+		 + I/O latency.
+* mmp_eck:       Embedded checksum of the MMP data.  To work with existing
+		 zio code, this will be stored at the end of the MMP block,
+		 and is not part of the MMP block itself.
+
+Suggested fields:
+
+* mmp_nodename:  Hostname of the node that wrote this MMP block, for admin
+                 notification purposes.
+
+All nodes must know reliably where to read this data, since these on-disk
+structures are the communication medium between nodes.  This is simplest if the
+data is written to a fixed on-disk position.
```

Comment:
For anyone else following this conversation, the answer is: "slots in the uberblock rings within the labels, on leaf vdevs"